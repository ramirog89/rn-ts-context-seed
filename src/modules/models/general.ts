export type IAction<T> = { type: T | null, payload?: any }

export type Nullable<T> = { [P in keyof T]: T[P] | null };

export type IEntityMap<T> = { [key:string]: T };

export type ILoading = { isLoading: boolean; error?: any; hasError?: boolean; };

export enum ToastType {
  SUCCESS = 'success',
  ERROR = 'error',
  WARNING = 'warning'
}

export interface IToast {
  message: string;
  type: ToastType;
};
