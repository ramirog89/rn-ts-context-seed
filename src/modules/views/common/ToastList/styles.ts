import { StyleSheet } from 'react-native';

export const styles: any = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 25,
    display: 'flex',
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    zIndex: 10
  },
  toast: {
    width: '50%',
    borderColor: 'gray',
    borderWidth: 1,
    borderRadius: 5,
    padding: 5,
    color: 'white'
  },
  success: {
    backgroundColor: 'green'
  },
  warning: {
    backgroundColor: 'yellow'
  },
  error: {
    backgroundColor: 'red'
  }
});
