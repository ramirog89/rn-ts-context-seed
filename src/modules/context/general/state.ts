import { GeneralModel } from '../../models';

export interface IState {
  loading: GeneralModel.IEntityMap<GeneralModel.ILoading>;
  toastList: { id: string, toast: GeneralModel.IToast }[];
};

export const initialState: IState = {
  loading: {},
  toastList: []
}
