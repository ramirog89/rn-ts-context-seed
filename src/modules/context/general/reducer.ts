import { GeneralModel } from '../../models';
import { ActionType } from './actions';
import { IState } from './state';

export const reducer = (state: IState, action: GeneralModel.IAction<ActionType>) => {
  switch (action.type) {
    case ActionType.ADD_TOAST:
      return { ...state, toastList: [...state.toastList, action.payload] };
    case ActionType.REMOVE_TOAST:
      return { ...state, toastList: state.toastList.filter((toast) => action.payload.id !== toast.id) };
    default:
      return state;
  }
};

