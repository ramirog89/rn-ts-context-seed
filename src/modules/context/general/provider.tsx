import React, { useMemo, createContext, useReducer, useContext, useEffect } from 'react';

import { wait } from '../../../utils/wait';
import { GeneralModel } from '../../models';
import { IProviderProps } from '../rootState';
import { IState, initialState } from './state';
import { ActionType } from './actions';
import { reducer } from './reducer';

export const GeneralContext = createContext<
  {
    state: IState;
    addToast: (toast: GeneralModel.IToast) => void;
  } | null
>(null);

export const useGeneral = () => {
  const context = useContext(GeneralContext);
  if (!context) {
    throw new Error('GeneralContext is not wrapped with GeneralContext.Provider');
  }
  return context;
};

export const GeneralProvider = (props: IProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addToast = async (toast: GeneralModel.IToast) => {
    const id = Date.now().toString();
    dispatch({ type: ActionType.ADD_TOAST, payload: { id, toast } });
    await wait(2000);
    dispatch({ type: ActionType.REMOVE_TOAST, payload: { id } });
  };

  const value = useMemo(() => ({ state, addToast }), [state]);

  return <GeneralContext.Provider value={value} {...props} />
};

