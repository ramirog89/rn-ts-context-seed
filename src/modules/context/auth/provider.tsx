import React, { useMemo, createContext, useReducer, useContext, useCallback } from 'react';
import jwt_decode from 'jwt-decode';

import { GeneralModel, UserModel } from '../../models';
import { IProviderProps } from '../rootState';
import { IState, initialState } from './state';
import { ActionType } from './actions';
import { reducer } from './reducer';
import { useGeneral } from '../general';

interface IAuthContext {
  state: IState;
  signIn: (credentials: UserModel.ILoginRequest) => void;
  signOut: () => void;
  recoverSession: () => void;
}

export const AuthContext = createContext<IAuthContext | null>(null);

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error('AuthContext is not wrapped with AuthContext.Provider');
  }
  return context;
};

export const AuthProvider = (props: IProviderProps) => {
  const { deps } = props;
  const [state, dispatch] = useReducer(reducer, initialState);
  const { addToast } = useGeneral();

  const signIn = useCallback(async  (credentials: UserModel.ILoginRequest) => {
    try {
      const response = await deps.apiService.signIn(credentials);
      const user = jwt_decode(response.token);
      await deps.storageService.set('token', response.token);
      dispatch({ type: ActionType.AUTH_SIGN_IN_SUCCESS, payload: { user } });
    } catch (e) {
      addToast({ message: 'authSignIn error', type: GeneralModel.ToastType.ERROR });
    }
  }, [dispatch]);

  const recoverSession = useCallback(async () => {
    const token: string | undefined = await deps.storageService.get('token');
    if (token) {
      const user = jwt_decode(token);
      dispatch({ type: ActionType.AUTH_SIGN_IN_SUCCESS, payload: { user } });
    } else {
      dispatch({ type: ActionType.AUTH_SIGN_OUT, payload: {} });
    }
  }, []);

  const signOut = useCallback(async () => {
    try {
      await deps.storageService.remove('token');
      dispatch({ type: ActionType.AUTH_SIGN_OUT, payload: {} });
    } catch (e) {
      addToast({ message: 'authSignOut error', type: GeneralModel.ToastType.ERROR });
    }
  }, [dispatch]);

  const value = useMemo(() => ({ state, signIn, signOut, recoverSession }), [state]);

  return <AuthContext.Provider value={value} {...props} />
};

