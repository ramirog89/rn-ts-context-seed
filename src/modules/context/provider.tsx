import React from 'react';
import RNSecureKeyStore from 'react-native-secure-key-store';

import { ApiService } from '../services/ApiService';
import { StorageService } from '../services/StorageService';

import { ComposeProvider, Provider } from './compose';
import { IDependencies } from './rootState';
import { GeneralProvider } from './general/provider';
import { AuthProvider } from './auth/provider';

export const dependencies: IDependencies = {
  apiService: new ApiService(),
  storageService: new StorageService({ storage: RNSecureKeyStore })
};

export const providers: Provider[] = [GeneralProvider, AuthProvider];

export const ContextProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ComposeProvider deps={dependencies} providers={providers}>
      {children}
    </ComposeProvider>
  )
}