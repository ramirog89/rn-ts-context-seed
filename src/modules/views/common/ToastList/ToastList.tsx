import React from 'react';
import { View, Text } from 'react-native';
import { useGeneral } from "../../../context/general/provider";

import { styles } from './styles';

const ToastList = () => {
  const { state } = useGeneral();

  return (
    <View style={styles.container}>
      {state.toastList.map((item) => (
        <View style={[styles.toast, styles?.[item.toast.type]]} key={item.id}>
          <Text>{item.toast.message}</Text>
        </View>
      ))}
    </View>
  );
}

export default ToastList;
