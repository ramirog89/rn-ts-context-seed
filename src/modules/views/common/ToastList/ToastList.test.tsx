import React from 'react';

import { render } from '../../../../test/render';
import { getToast1 } from '../../../../test/entities';
import { GeneralContext } from '../../../context/general/provider';

import ToastList from './ToastList';

describe('ToastList', () => {
  it('should render', () => {
    const { toJSON } = render(
      <GeneralContext.Provider value={{ state: { loading: {}, toastList: [{ id: '1', toast: getToast1() }], }, addToast: jest.fn() }}>
        <ToastList />
      </GeneralContext.Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
