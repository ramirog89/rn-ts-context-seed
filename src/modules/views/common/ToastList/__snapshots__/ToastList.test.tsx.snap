// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastList should render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "center",
      "position": "absolute",
      "top": 10,
      "width": "100%",
      "zIndex": 10,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 5,
          "display": "flex",
          "justifyContent": "center",
          "marginBottom": 5,
          "opacity": 0.9,
          "padding": 10,
          "width": "80%",
        },
        Object {
          "backgroundColor": "green",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
        }
      }
    >
      tosat
    </Text>
  </View>
</View>
`;
